Index: gitlab/app/assets/javascripts/confidential_merge_request/components/project_form_group.vue
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- gitlab/app/assets/javascripts/confidential_merge_request/components/project_form_group.vue	(date 1574672573000)
+++ gitlab/app/assets/javascripts/confidential_merge_request/components/project_form_group.vue	(date 1574859869708)
@@ -1,6 +1,6 @@
 <script>
-import { GlLink } from '@gitlab/ui';
-import { __, sprintf } from '../../locale';
+import { GlLink, GlSprintf } from '@gitlab/ui';
+import { __ } from '../../locale';
 import createFlash from '../../flash';
 import Api from '../../api';
 import state from '../state';
@@ -9,6 +9,7 @@
 export default {
   components: {
     GlLink,
+    GlSprintf,
     Dropdown,
   },
   props: {
@@ -38,18 +39,6 @@
     selectedProject() {
       return state.selectedProject;
     },
-    noForkText() {
-      return sprintf(
-        __(
-          "To protect this issue's confidentiality, %{link_start}fork this project%{link_end} and set the fork's visibility to private.",
-        ),
-        {
-          link_start: `<a href="${this.newForkPath}" target="_blank" class="help-link">`,
-          link_end: '</a>',
-        },
-        false,
-      );
-    },
   },
   mounted() {
     this.fetchProjects();
@@ -127,7 +116,17 @@
         </template>
         <template v-else>
           {{ __('No forks are available to you.') }}<br />
-          <span v-html="noForkText"></span>
+          <gl-sprintf
+            :message="
+              __(
+                `To protect this issue's confidentiality, %{forkLink} and set the fork's visibility to private.`,
+              )
+            "
+          >
+            <template #forkLink>
+              <a :href="newForkPath" target="_blank" class="help-link">fork this project</a>
+            </template>
+          </gl-sprintf>
         </template>
         <gl-link
           :href="helpPagePath"
